langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/ToolExecutionHelper.java

```
@@ -21,24 +20,24 @@ static String extractResult(JsonNode result) {
             JsonNode resultNode = result.get("result");
             if (resultNode.has("content")) {
                 String content = extractSuccessfulResult((ArrayNode) resultNode.get("content"));
-                boolean isError = false;
-                if (resultNode.has("isError")) {
-                    isError = resultNode.get("isError").asBoolean();
-                }
-                if (isError) {
-                    content = String.format(EXECUTION_ERROR_MESSAGE + ". The tool returned: %s", content);
+                if (isError(resultNode)) {
+                    throw new ToolExecutionException(content);
```

@jmartisk I am trying to make `DefaultToolExecutor` and `McpToolExecutor` to behave consistently (in this PR):
- when it cannot parse tool arguments - throw `ToolArgumentParsingException`
- when it fails to execute the tool - throw `ToolExecutionException`.

Because of this, I need to adapt the behaviour of `ToolExecutionHelper` and `DefaultMcpClient`, do you see any problems with that? It is a breaking change, but since MCP module is still in beta, I guess it should not be a big deal?

---

@jmartisk polite ping :)

---

Oh, sorry, I'd completely missed this! Feel free to prod me on Slack also whenever this happens.

> when it cannot parse tool arguments - throw ToolArgumentParsingException 

OK

> when it fails to execute the tool - throw ToolExecutionException. 

So, don't you need to also introduce a `propagateToolExecutionExceptions` field for MCP clients too? IIUC with your current code, it would always throw an exception instead of returning something to the chat model?

---

One might want to return an error string instead (in some cases, it may be expected that an MCP tool throws an error)

---

Or even better, we could allow both `DefaultToolExecutor` and `McpToolExecutor` to have a custom `ToolErrorHandler` class that dynamically decides how the error would be handled. But maybe that would be unnecessarily complex...

---

> Or even better, we could allow both DefaultToolExecutor and McpToolExecutor to have a custom ToolErrorHandler class that dynamically decides how the error would be handled. But maybe that would be unnecessarily complex...

That is what this PR is trying to achieve :)

The idea is that `ToolExecutor` throws an exception when error happens (instead of returning a `String`, which hides the fact that error happened), and depending on the error type, either `ToolArgumentsErrorHandler` or `ToolExecutionErrorHandler` will handle it.

> So, don't you need to also introduce a propagateToolExecutionExceptions field for MCP clients too? IIUC with your current code, it would always throw an exception instead of returning something to the chat model?

I had to introduce `propagateToolExecutionExceptions` in `DefaultToolExecutor` because that module is stable and we cannot break the default behaviour. It adds more complexity, but keeps behaviour unchanged for those who interact with `DefaultToolExecutor` directly.

`McpToolExecutor`, on the other hand, is still beta/experimental, so we could break it IMHO. Also, I am not sure if many users really call `DefaultMcpClient.executeTool()` or `McpToolExecutor.execute()` directly. Only those who call them directly will see the breaking changes, those who configure `McpToolProvider` on AI Service will not see any changes.

---

If you look into `ToolService`, there are `DEFAULT_TOOL_ARGUMENTS_ERROR_HANDLER` and `DEFAULT_TOOL_EXECUTION_ERROR_HANDLER` which keep the default behaviour for tool execution errors (both default and MCP) untouched, only the "API contract" between AI Service and `ToolExecutor` is changing: instead of returning an error as a `String` from `ToolExecutor`, an exception is thrown. This way we can detect in AI Service that error happened and call the appropriate error handler. Or, potentially in the future, do something else.

---

Ah ok, thanks for explaining (I was too lazy to read the whole code), it makes sense to me, thanks!